src/firmware/shared/circular_buffer.c

```diff
@@ -0,0 +1,63 @@
+#include "firmware/shared/circular_buffer.h"
+
+#include <assert.h>
+
+struct CircularBuffer
+{
+    int head;
+    int tail;
+    int size;
+    float *bufferData;
+    bool isFull;
+};
+
+CircularBuffer_t *circular_buffer_create(CircularBuffer_t *cbuffer, int size)
+{
+    CircularBuffer_t *new_cbuffer = malloc(sizeof(CircularBuffer_t));
+
+    new_cbuffer->head       = 0;
+    new_cbuffer->tail       = 0;
+    new_cbuffer->size       = size;
+    new_cbuffer->bufferData = (float *)malloc(sizeof(float) * size);
+    new_cbuffer->isFull     = false;
+
+    return new_cbuffer;
+}
+
+void circular_buffer_free(CircularBuffer_t *cbuffer)
+{
+    free(cbuffer);
```

Comment:
I think we'll leak `bufferData` here?

Reply:
Added a `free(cbuffer->bufferData)` before `free(cbuffer)`.